Adding the transfer_to_output function to token library

nfurfaro opened the pull request:
The 4th and final(for now) function for the token library.
>/// Transfer `amount` coins of type `asset_id` to address `recipient`.

The comments in the new function should provide a fairly clear explanation of what is going on.

Tests here: https://github.com/FuelLabs/sway/pull/722

adlerjohn commented on src/token.sw:
```diff
@@ -25,3 +27,66 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
+// TransactionScript outputsCount has a 48 byte(6 words * 8 bytes) offset
+// Transaction Script: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#transactionscript
+// Output types: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#output
+// const OUTPUT_LENGTH_LOCATION = 48;
+// const OUTPUT_VARIABLE_TYPE = 4;
```
Why don't you use these constants?

nfurfaro replied:
Ah. I had disabled them while trying to pin down the source of the non-deterministic contract-ids issue.
They turned out to not be the cause, but I forgot to use them again. 
Will fix.

adlerjohn reviewed (approved):
utACK

otrho commented on src/token.sw:
```diff
@@ -25,3 +27,66 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
+            xos type slot;
+            meq res type target bytes;
+            res: bool
+        };
```
I don't think the above is doing what you want it to.

`xos type slot`  will get the offset to the `slot`th output value and put it in `type`.

`meq res type target bytes`  will compare the two memory regions at `type` and `target` of `bytes` (8) bytes and put it in `res`.  `type` is the offset to the output value, but `target` is 4 -- the value you want to compare against, not an offset.  Since it's a number, and not a struct, array or anything larger than a word, then `target` will be passed to the asm block by value.  I might be wrong, but this is my understanding.

And also according to that spec, the output type is a single byte, so you shouldn't be comparing 8 bytes.

What I think you want is to call `xos` and then just `lb` on what it returns to read the first byte in the output, which is the type.  Then compare it to `OUTPUT_VARIABLE_TYPE` outside the asm block.

nfurfaro replied:
fixed.

otrho commented on src/token.sw:
```diff
@@ -25,3 +27,66 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
+                meq is_zero a zero bytes;
+                is_zero: bool
+            };
+            // && if the amount is zero:
```
The spec says that for a variable output the amount is the second field, and the first field `to` is a `byte[32]` so you should be adding only `i32` to `output` to get `amount_ptr`, as well as that type byte, so I'm thinking the offset should be `addi amount_ptr output i33`.

This time you're using `lw` which I hope will work from unaligned offsets @adlerjohn?

But then instead of `meq` you should just use `eq`, to compare values, or preferably do the comparison outside the asm block.

adlerjohn replied:
I think adding `i32` should be sufficient, since `lw` would start from the leftmost byte.

nfurfaro replied:
I will fix the offset to `i32` and extract the comparison out of the `asm` block.

otrho replied:
Hang on.  `XOS` returns the offset to the output, which has a `type` and `data`.  The `type` is a `uint8` but if everything is word aligned then it uses 8 bytes.  Then there's the variable data, `to`, `amount` and `asset_id`.  So the offset from what `XOS` returns and the amount is 8+32=40.

otrho replied:
Also, using `target_output_type_exists` from above and `amount_is_zero` here is a little weird -- you can just put those comparisons into the `if` condition.   `if type == OUTPUT_VARIABLE_TYPE { ... }` and `if amount == 0 { ... }`.

nfurfaro replied:
re: "using target_output_type_exists from above "
yep, that was an oversight on my part. 
I've removed both redundant variables and made the check for equality in the `if` conditions directly.

otrho commented on src/token.sw:
```diff
@@ -25,3 +27,66 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
+    // TransactionScript outputsCount has a 48 byte(6 words * 8 bytes) offset
+    // Transaction Script: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#transactionscript
+    // Output types: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#output
+    const OUTPUT_LENGTH_LOCATION = 48;
```
I don't think 48 is right.  If I'm looking at the right table, there's `gasPrice`, `gasLimit`, `bytePrice` are 8 byte words, then `maturity` is a 4 byte word, then `scriptLength` and `scriptDataLength` are 2 byte words, and then `inputsCount` and `outputsCount` are bytes.

8+8+8+4+2+2+1 is 33.

adlerjohn replied:
Everything is word aligned, so things less than a word would be aligned to a word. So it would be 8+8+8+8+8+8+8 = 56, since an extra field was added recently. Used to be 48.

otrho commented on src/token.sw:
```diff
@@ -25,3 +27,66 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
+    // get length of outputs from TransactionScript outputsCount:
+    let length: u8 = asm(outputs_length, outputs_length_ptr: OUTPUT_LENGTH_LOCATION) {
+        lw outputs_length outputs_length_ptr i0;
+        outputs_length: u8
```
Since the `outputsCount` value is a byte, you want to use `lb` to load it, rather than `lw` which will load 7 extra bytes from the header too.

adlerjohn replied:
You could do either since things are word-aligned.

otrho replied:
OK, I missed that detail.  Why are they documented as smaller types when they're all aligned?  They might as well be all `uint64`s.

sezna replied:
`lb` is _more_ correct though, I think, since it isn't actually part of the spec (or at least a part we'd want to rely on) that the rest of the bytes will be zeroed out and/or ignored.

nfurfaro replied:
I replaced `lw` with `lb`.

nfurfaro commented:
I've addressed the comments above.
Note that as per https://github.com/FuelLabs/sway/issues/685 , tests for this code (which involve contract calls) always pass even when they shouldn't. I think we should wait to merge this until we actually have working test coverage for it.

otrho reviewed (changes requested):
Needing clarification on the offset to the output variable amount field.

nfurfaro commented:
>Needing clarification on the offset to the output variable amount field.

Me too. 
What I need to know is, for the `OutputVariable`'s `to` field (which is a 32-byte value), does `to` take up 4 words (32/8) or only a single word with a pointer to the actual 32-byte value?
If it's the former, then you're right @otrho , the offset from what `xos` returns to the `value` is actually 40 bytes. 
If it's the latter, then it looks like it would only be a 16-byte offset.
@adlerjohn

nfurfaro commented on src/token.sw:
```diff
@@ -25,3 +27,72 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
+    // If an output of type `OutputVariable` is found, check if its `amount` is zero.
+    // As one cannot transfer zero coins to an output without a panic, a variable output with a value of zero is by definition unused.
+    while index < length {
+        let output_start = asm(n: index, offset) {
```
Note: I separated the use of the `xos` opcode into its own asm block here to avoid calling it twice per iteration.
this way, `output_start` can be reused below, passing it into both the `type` and `amount` asm blocks

adlerjohn commented:
> What I need to know is, for the OutputVariable's to field (which is a 32-byte value), does to take up 4 words (32/8) or only a single word with a pointer to the actual 32-byte value?

It should be the 32-byte hash directly, no pointer. So 4 words.

adlerjohn reviewed (approved):
utACK
